src/python/pants/core/target_types.py

```diff
@@ -206,5 +216,142 @@ class GenericTarget(Target):
     core_fields = (*COMMON_TARGET_FIELDS, Dependencies)
 
 
+# -----------------------------------------------------------------------------------------------
+# `archive` target
+# -----------------------------------------------------------------------------------------------
+
+# TODO(#10888): Teach project introspection goals that this is a special type of the `Dependencies`
+#  field.
+class ArchivePackages(StringSequenceField):
+    """Addresses to any targets that can be built with `./pants package`.
+
+    Pants will build the assets as if you had run `./pants package`. It will include the
+    results in your archive using the same name they would normally have, but without the `dist/`
+    prefix.
+
+    You can include anything that can be built by `./pants package`, e.g. a `python_binary`,
+    `python_awslambda`, or even another `archive`.
+    """
+
+    alias = "packages"
+
+
+# TODO(#10888): Teach project introspection goals that this is a special type of the `Dependencies`
+#  field.
+class ArchiveFiles(StringSequenceField):
+    """Addresses to any `files` or `relocated_files` targets to include in the archive, e.g.
+    `["resources:logo"]`.
+
+    This is useful to include any loose files, like data files, image assets, or config files.
+
+    This will ignore any targets that are not `files` or `relocated_files` targets. If you instead
+    want those files included in any packages specified in the `packages` field for this target,
+    then use a `resources` target and have the original package (e.g. the `python_library`)
+    depend on the resources.
+    """
+
+    alias = "files"
+
+
+class ArchiveFormatField(StringField):
+    """The type of archive file to be generated.
+
+    Note that `zip` files will not be compressed.
+    """
+
+    alias = "format"
+    valid_choices = ArchiveFormat
+    required = True
+    value: str
+
+
+class ArchiveTarget(Target):
+    """An archive (e.g. zip file) containing loose files and/or packages built via `./pants
+    package`."""
+
+    alias = "archive"
+    core_fields = (*COMMON_TARGET_FIELDS, ArchivePackages, ArchiveFiles, ArchiveFormatField)
+
+
+@dataclass(frozen=True)
+class ArchiveFieldSet(PackageFieldSet):
+    required_fields = (ArchiveFormatField,)
+
+    packages: ArchivePackages
+    files: ArchiveFiles
+    format_field: ArchiveFormatField
+
+
+@rule(level=LogLevel.DEBUG)
+async def package_archive_target(field_set: ArchiveFieldSet) -> BuiltPackage:
+    package_targets = await MultiGet(
+        Get(
+            WrappedTarget,
+            AddressInput,
+            AddressInput.parse(v, relative_to=field_set.address.spec_path),
+        )
+        for v in field_set.packages.value or ()
+    )
+    package_field_sets_per_target = await Get(
```

Comment:
Particularly if both the files and archives arguments are merged into one field, there should be some more parallelism available here... it's roughly:
1. get all `WrappedTargets` for all dependencies in parallel
2. get either a `BuiltPackage` or a `HydratedSources` for each of them in parallel